src/primitives/hrp.rs

```diff
@@ -0,0 +1,180 @@
+// Written by Clark Moody and the rust-bitcoin developers.
+// SPDX-License-Identifier: MIT
+
+//! Provides an `Hrp` type that represents the human-readable part of a bech32 encoded string.
+//!
+//! > The human-readable part, which is intended to convey the type of data, or anything else that
+//! > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+//! > character having a value in the range [33-126]. HRP validity may be further restricted by
+//! > specific applications.
+//!
+//! ref: [BIP-173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32)
+
+#[cfg(feature = "alloc")]
+use alloc::borrow::Cow;
+use core::fmt;
+use core::iter::FusedIterator;
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Hrp<'a> {
+    /// Lowercase human-readable part.
+    Lower(&'a str),
+    /// Uppercase human-readable part.
+    Upper(&'a str),
+    /// All digit human-readable part.
+    Digit(&'a str),
+}
+
+impl<'s> Hrp<'s> {
+    /// Parses the human-readable part checking it is valid as defined by [BIP-173].
+    ///
+    /// This does _not_ check that the `hrp` is an in-use HRP within Bitcoin (eg, "bc"), rather it
+    /// checks that the HRP string is valid as per the specification in [BIP-173]:
+    ///
+    /// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+    pub fn parse(hrp: &'s str) -> Result<Self, Error> {
+        use Error::*;
+
+        if hrp.is_empty() {
+            return Err(Empty);
+        }
+        if hrp.len() > MAX_HRP_LEN {
+            return Err(TooLong(hrp.len()));
+        }
+
+        let mut has_lower: bool = false;
+        let mut has_upper: bool = false;
+        for b in hrp.bytes() {
+            // Valid subset of ASCII
+            if !(33..=126).contains(&b) {
+                return Err(InvalidAsciiByte(b));
+            }
+
+            if b.is_ascii_lowercase() {
+                has_lower = true;
+            } else if b.is_ascii_uppercase() {
+                has_upper = true;
+            };
+
+            if has_lower && has_upper {
+                return Err(MixedCase);
+            }
+        }
+
+        if has_lower {
+            return Ok(Hrp::Lower(hrp));
+        }
+
+        if has_upper {
+            return Ok(Hrp::Upper(hrp));
+        }
+
+        Ok(Hrp::Digit(hrp))
+    }
+
+    /// Returns the human-readable part as a lowercase string.
+    #[cfg(feature = "alloc")]
+    pub fn to_lower(&self) -> Cow<str> {
+        match self {
+            Self::Upper(upper) => Cow::Owned(upper.to_lowercase()),
+            Self::Lower(lower) => Cow::Borrowed(lower),
+            Self::Digit(digits) => Cow::Borrowed(digits),
```

Comment:
I guess this would be less efficient if we weren't tracking case, but I don't really see the use of this method anyway, so I think that's fine.